R/toolCalcEnergyProj.R

```diff
@@ -0,0 +1,142 @@
+    } else if (subtype == "SE|Electricity") {
+      xHist <- collapseDim(IEA[, , "SE|Electricity|Biomass (EJ/yr)"] + IEA[, , "SE|Electricity|Solar (EJ/yr)"] + IEA[, , "SE|Electricity|Wind (EJ/yr)"] +
+                             IEA[, , "SE|Electricity|Hydro (EJ/yr)"] + IEA[, , "SE|Electricity|Geothermal (EJ/yr)"] + IEA[, , "SE|Electricity|Coal (EJ/yr)"] + IEA[, , "SE|Electricity|Gas (EJ/yr)"] + IEA[, , "SE|Electricity|Oil (EJ/yr)"] +
+                             IEA[, , "SE|Electricity|Nuclear (EJ/yr)"], dim = 3)
+    }
+
+    if (AvgSeveralYears) {
+      AvgYears <- c("y2018", "y2019", "y2020", "y2021", "y2022")
+      xHist <- collapseDim(dimSums(xHist[, AvgYears, ], dim = 2) / length(AvgYears))
+    } else {
+      # only get 2020 year
+      xHist <- collapseDim(xHist[, "y2020", ])
+    }
+
+    return(xHist)
+  }
+
+  # function to calculate FE trend on level of REMIND regions based on EDGE data
+  calcFeTrend <- function(subtype, subset, years) {
+    # get FE demand projections per REMIND region for transport, buildings and industry for SSP2
+    # transport FE demand from EDGE-T (in TWa/yr)
+    FETransport <- calcOutput("EDGETransport", subtype = "f35_demByTech")
+    # filter for SSP2 scenario and Mix1 EDGE scenario
+    FETransport <- collapseNames(mselect(FETransport, GDP_scenario = subset, DEM_scenario = subset, EDGE_scenario = "Mix1"))
+    # buildings and industry FE demand from EDGE-B and EDGE-I (in EJ/yr)
+    FEBuildIndustry <- calcOutput("FEdemand", scenario = subset, signif = 4)
+    # filter for SSP2 scenario
+    FEBuildIndustry <- collapseNames(FEBuildIndustry[, , subset])
+
+    # define CES inputs over which to sum for respective FE demand
+    if (subtype == "FE") {
+      # for FE projection, get FE trend of total FE
+      input.fe.build <- c("feelcb", "feelhpb", "feelrhb", "fegab", "fehob", "fesob", "feheb")
+      input.fe.indst <- c(
+        grep("fe.*steel", getNames(FEBuildIndustry), value = T),
+        grep("fe.*cement", getNames(FEBuildIndustry), value = T),
+        grep("fe.*chem", getNames(FEBuildIndustry), value = T),
+        grep("fe.*otherInd", getNames(FEBuildIndustry), value = T)
+      )
+      input.fe.trans <- getNames(FETransport)
+    } else if (subtype == "SE|Electricity") {
+      # for SE electricity projection, get FE trend of FE electricity
+      input.fe.build <- c("feelcb", "feelhpb", "feelrhb")
+      input.fe.indst <- c(
+        "feel_steel_secondary", "feel_steel_primary", "feel_cement",
+        "feelwlth_chemicals", "feelhth_chemicals",
+        "feelwlth_otherInd", "feelhth_otherInd"
+      )
+      input.fe.trans <- c("feelt")
+    }
+
+    # common years across EDGE outputs
+    common.years <- intersect(getYears(FEBuildIndustry), getYears(FETransport))
+
+    # sum FE demand across sectors per REMIND region
+    FeTotal <- dimSums(FEBuildIndustry[, common.years, input.fe.build], dim = 3) +
```

Comment:
Are you sure that you cannot get the total FE demand another way. You calculation is expensive and not robust in terms of caching due to the sources used. Besides, is seems redundant to me, as I would expect that we already have madrat functions that calculate this?